File: DFS_BFS_2.cpp
//
// Created by Macbook Pro on 2020/12/06.
//

#include <iostream>
#include <vector>
using namespace std;

void dfs(int n, vector<vector<int>>& computers){

    if(!computers[n][n]) return;
    computers[n][n] = 0;
    for(int i=0;i<computers.size();i++){
        if(computers[n][i] && computers[i][i]){
            dfs(i,computers);
        }
    }
}

int solution(int n, vector<vector<int>> computers) {
    int answer=0;

    for(int i=0;i<n;i++){
        if(!computers[i][i]) continue;
        answer++;
        dfs(i,computers);
    }

    return answer;
}
int main(){
    int n = 5;
    vector<vector<int>> computers = {{1,0,1,1,0},
                                     {0,1,0,1,0},
                                     {1,0,1,0,0},
                                     {1,1,0,1,1},
                                     {0,0,0,1,1}};
    cout << solution(n,computers) << endl;
}

File: python/programmers/dfs_bfs_2.py
# https://programmers.co.kr/learn/courses/30/lessons/43162

from collections import defaultdict

def dfs(comuters, check, index):
    check[index] = True

    for i in range(len(check)):
        if computers[index][i] and not check[i]:
            dfs(comuters, check, i)



def solution(computers, n):
    answer = 0
    check = [False] * n

    for i in range(n):
        if not check[i]:
            dfs(computers, check, i)
            answer+=1

    return answer


if __name__ == "__main__":
    # computers = [[1, 1, 0], [1, 1, 0], [0, 0, 1]]
    computers = [[1, 1, 0], [1, 1, 1], [0, 1, 1]]
    n = 3
    print(solution(computers, n))